### PR TITLE
[FLINK-18512][kinesis] Introducing RecordPublisher. Refactor ShardConsumer to use PollingRecordPublisher

### PR DESCRIPTION
**Pull request by dannycranmer**
## What is the purpose of the change

This is the second milestone of [FLIP-128](https://cwiki.apache.org/confluence/display/FLINK/FLIP-128%3A+Enhanced+Fan+Out+for+AWS+Kinesis+Consumers) to add EFO support to the `FlinkKinesisConsumer`. The consumption mechanism in `ShardConsumer` has been abstracted to an interface and implemented to retain existing behaviour. An EFO implementation will be added in a future contribution. This change should not introduce any functional differences.

## Note
This PR is blocked by:
- Test coverage improvements:
  - [JIRA issue](https://issues.apache.org/jira/browse/FLINK-18483)
  - [Pull request](https://github.com/apache/flink/pull/12850)

## Brief change log

- `RecordPublisher`
  - Added a new interface used to consume records from Kinesis and supply them to Flink
- `PollingRecordPublisher` 
  - The existing consumption mechanism using `AWS::GetRecords` and `AWS::GetShardIterator` has been refactored into an implementation of `RecordPublisher`. A factory has been added to create instances
  - Configuration that was deserialised in the constructor has been moved to a class `PollingRecordPublisherConfiguration`
- `AdaptivePollingRecordPublisher`
  - When a user supplies `useAdaptiveReads` a special `RecordPublisher` is created that adds adaptive loop delay and batch size. This functionality has been moved from `ShardConsumer`, but now has it's own implementation that extends `PollingRecordPublisher`
-  `StartingPosition`
   - A new object has been created to encapsulate starting sequence number/iterator type. A starting position can be created from a `SentinelSequenceNumber` 
- `ShardMetricsReporter`
  - Metric reporting has been split to allow arbitrary metrics to be registered by `RecordPublisher` implementations
  - `ShardConsumer` reports general metrics using `ShardConsumerMetricsReporter`
  - `PollingRecordPublisher` reports metrics using `PollingRecordPublisherMetricsReporter`
- ` ShardConsumer`
  - Refactored to accept a `RecordPublisher` and not depend on a `KinesisProxyInterface`
  - Restarting from aggregated sequence number logic moved to support generic `RecordPublishers`:
    - Before: At startup a single record was downloaded and already processed subsequences were discarded
    - After: During consumption any already processed subsequences are discarded
    - Note: This does add a small overhead to the continuous consumption, but simplifies `RecordPublishers`. An alternative could be to code the `ShardConsumer` to handle restarting from aggregated records via a standard `KinesisProxy`

## Verifying this change

This change is already covered by existing tests, such as:
- `ShardConsumerTest`

A preceding pull request was submitted to increase test coverage before performing the refactor:
- https://github.com/apache/flink/pull/12850

This change added tests and can be verified as follows:
- `RecordBatch` -> `RecordBatchTest`
- `PollingRecordPublisherConfiguration` -> `PollingRecordPublisherConfigurationTest`
- `PollingRecordPublisherFactory` -> `PollingRecordPublisherFactoryTest`
- `PollingRecordPublisher` -> `PollingRecordPublisherTest`
- `PollingRecordPublisherMetricsReporter` -> `PollingRecordPublisherMetricsReporterTest`
- `ShardConsumerMetricsReporter` -> `ShardConsumerMetricsReporterTest`
- `StartingPosition` -> `StartingPositionTest`

I have deployed Flink applications locally and on AWS KDA-Java and verified consumption.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no (dependency change is covered by FLINK-18483)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes (added check for filtering de-aggregated subsequence numbers `ShardConsumer::filterDeaggregatedRecord`)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes (restarting consumption from sequence number)
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by xiaolong-sn** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/model/StartingPosition.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
+/*
```
Comment (xiaolong-sn) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/model/StartingPosition.java:1]: This class looks quite like the `SentinelSequenceNumber ` class, is there any possibility to combine those two classes?

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 2]
```diff
@@ -379,49 +205,18 @@ private void deserializeRecordForCollectionAndUpdateState(UserRecord record)
...
-
-				shardItr = getShardIterator(lastSequenceNum);
-
-				// sleep for the fetch interval before the next getRecords attempt with the refreshed iterator
-				if (fetchIntervalMillis != 0) {
-					Thread.sleep(fetchIntervalMillis);
-				}
-			}
+	private boolean filterDeaggregatedRecord(final UserRecord record) {
+		if (lastSequenceNum.isAggregated()) {
```
[the remark this answers is not in the document]
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java:215]: It is replicating this behaviour: https://github.com/apache/flink/blob/release-1.10.1/flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java#L198

Say you have an aggregated message X with 6 child records and you stop consuming at subsequence 3, the last consumed sequence number would be:
- sequence: X
- sub-sequence: 3

Therefore when you restart at X, you need to discard subsequence 1,2,3 and process 4,5,6.

This method discards any previously processed subsequence numbers continuously, rather than upfront. This allows the logic to be pulled into the `ShardConsumer` from the `RecordPublisher` and support EFO/Polling without configuring multiple clients. As explained in the PR description, the code is simpler putting this check here at the expense of a slight overhead.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyMarker.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package org.apache.flink.streaming.connectors.kinesis.proxy;
+
+import org.apache.flink.annotation.Internal;
+
+/**
+ * A marker interface for generic Kinesis Proxy.
+ */
+@Internal
```
[the remark this answers is not in the document]
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyMarker.java:25]: It is simply a marker interface for generics on the `RecordPublisherFactory`, used to denote a `KinesisProxy`, be that SDK v1 or v2.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/PollingRecordPublisher.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	 * @param startingPosition the position in the stream in which to consume from
+	 * @param maxNumberOfRecords the maximum number of records to fetch for this getRecords attempt
+	 * @return get records result
+	 */
+	private GetRecordsResult getRecords(String shardItr, final StartingPosition startingPosition, int maxNumberOfRecords) throws InterruptedException {
+		GetRecordsResult getRecordsResult = null;
+		while (getRecordsResult == null) {
+			try {
+				getRecordsResult = kinesisProxy.getRecords(shardItr, maxNumberOfRecords);
+			} catch (ExpiredIteratorException | InterruptedException eiEx) {
```
[the remark this answers is not in the document]
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/PollingRecordPublisher.java:131]: Ah that `InterruptedException` should not be there! I will delete that

The other exceptions are handled in the proxy: https://github.com/apache/flink/blob/c5915cf87f96e1c7ebd84ad00f7eabade7e7fe37/flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java#L248

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/model/StartingPosition.java (hunk 1 above)
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/model/StartingPosition.java:1]: We cannot use `SentinelSequenceNumber` to achieve what this class does, as `SentinelSequenceNumber` is an enum. We could get rid of `SentinelSequenceNumber` in favour of this, however that class has a lot of references and would be a big refactor. Also I think semantically they differ.

Also looking at the code, `SentinelSequenceNumber` is potentially serialised and therefore we would need to migrate snapshots or handle deserialisation mappings which is risky.

**Comment by dannycranmer** [comment]
@tzulitai this pull request is ready for your review, thanks.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 5]
```diff
@@ -53,283 +49,111 @@
...
-					shardMetricsReporter.setNumberOfDeaggregatedRecords(numberOfDeaggregatedRecords);
-
-					nextShardItr = getRecordsResult.getNextShardIterator();
-
-					long adjustmentEndTimeNanos = adjustRunLoopFrequency(processingStartTimeNanos, System.nanoTime());
-					long runLoopTimeNanos = adjustmentEndTimeNanos - processingStartTimeNanos;
-					maxNumberOfRecordsPerFetch = adaptRecordsToRead(runLoopTimeNanos, fetchedRecords.size(), recordBatchSizeBytes, maxNumberOfRecordsPerFetch);
-					shardMetricsReporter.setRunLoopTimeNanos(runLoopTimeNanos);
-					processingStartTimeNanos = adjustmentEndTimeNanos; // for next time through the loop
+					startingPosition = StartingPosition.continueFromSequenceNumber(lastSequenceNum);
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java:145]: Why does the `ShardConsumer` need to keep track and update the `startingPosition`?
It feels like this should be an internal state to the record publisher, and also as how I understand it, the record publisher only ever respects the `nextShardItr` that it maintains.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 6]
```diff
@@ -379,49 +206,18 @@ private void deserializeRecordForCollectionAndUpdateState(UserRecord record)
...
+			return !record.getSequenceNumber().equals(lastSequenceNum.getSequenceNumber()) ||
+				record.getSubSequenceNumber() > lastSequenceNum.getSubSequenceNumber();
 		}
-		return getRecordsResult;
-	}
 
-	@SuppressWarnings("unchecked")
-	protected static List<UserRecord> deaggregateRecords(List<Record> records, String startingHashKey, String endingHashKey) {
-		return UserRecord.deaggregate(records, new BigInteger(startingHashKey), new BigInteger(endingHashKey));
+		return true;
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java:221]: Just for easier code readability, can we rewrite this as:

```
if (!lastSequenceNum.isAggregated()) {
    return true;
}

return !record.getSequenceNumber().equals(lastSequenceNum.getSequenceNumber()) ||
				record.getSubSequenceNumber() > lastSequenceNum.getSubSequenceNumber();
```

I've found early returns always helps with understanding these :)

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordBatch.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	private final int aggregatedRecordSize;
+
+	private final List<UserRecord> deaggregatedRecords;
+
+	private final long totalSizeInBytes;
+
+	private final Long millisBehindLatest;
+
+	public RecordBatch(final List<Record> records, final StreamShardHandle subscribedShard, final Long millisBehindLatest) {
+		this.aggregatedRecordSize = records.size();
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordBatch.java:46]: Argument sanity / precondition checks.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/AdaptivePollingRecordPublisher.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+import java.util.function.Consumer;
+
+/**
+ * An adaptive record publisher to add a dynamic loop delay and batch read size for {@link PollingRecordPublisher}.
+ * Kinesis Streams have quotas on the transactions per second, and throughout. This class attempts to balance
+ * quotas and mitigate back off errors.
+ */
+@Internal
+public class AdaptivePollingRecordPublisher extends PollingRecordPublisher {
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/AdaptivePollingRecordPublisher.java:36]: 👍 really like the approach you took in refactoring this.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/model/StartingPosition.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+import static com.amazonaws.services.kinesis.model.ShardIteratorType.AT_SEQUENCE_NUMBER;
+import static com.amazonaws.services.kinesis.model.ShardIteratorType.AT_TIMESTAMP;
+import static com.amazonaws.services.kinesis.model.ShardIteratorType.LATEST;
+import static com.amazonaws.services.kinesis.model.ShardIteratorType.TRIM_HORIZON;
+import static org.apache.flink.streaming.connectors.kinesis.model.SentinelSequenceNumber.isSentinelSequenceNumber;
+
+/**
+ * The position in which to start consuming from a stream.
+ */
+public class StartingPosition {
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/model/StartingPosition.java:36]: Mark with `@Internal`

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/AdaptivePollingRecordPublisher.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+	private long processingStartTimeNanos = System.nanoTime();
+
+	private int maxNumberOfRecordsPerFetch;
+
+	private final long fetchIntervalMillis;
+
+	private final PollingRecordPublisherMetricsReporter metricsReporter;
+
+	public AdaptivePollingRecordPublisher(
```
[the remark this answers is not in the document]
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/AdaptivePollingRecordPublisher.java]: Done

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/PollingRecordPublisher.java [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+	 * @param maxNumberOfRecordsPerFetch the maximum number of records to retrieve per batch
+	 * @param expiredIteratorBackoffMillis the duration to sleep in the event of an {@link ExpiredIteratorException}
+	 */
+	public PollingRecordPublisher(
+			final StreamShardHandle subscribedShard,
+			final PollingRecordPublisherMetricsReporter metricsReporter,
+			final KinesisProxyInterface kinesisProxy,
+			final int maxNumberOfRecordsPerFetch,
+			final long expiredIteratorBackoffMillis) {
+		this.subscribedShard = subscribedShard;
```
[the remark this answers is not in the document]
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/PollingRecordPublisher.java]: Done

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/PollingRecordPublisher.java [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+	 * A Polling implementation of {@link RecordPublisher} that polls kinesis for records.
+	 * The following KDS services are used: GetRecords and GetShardIterator.
+	 *
+	 * @param subscribedShard the shard in which to consume from
+	 * @param metricsReporter a metric reporter used to output metrics
+	 * @param kinesisProxy the proxy used to communicate with kinesis
+	 * @param maxNumberOfRecordsPerFetch the maximum number of records to retrieve per batch
+	 * @param expiredIteratorBackoffMillis the duration to sleep in the event of an {@link ExpiredIteratorException}
+	 */
+	public PollingRecordPublisher(
```
[the remark this answers is not in the document]
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/PollingRecordPublisher.java]: Done

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java (hunk 6 above)
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: Done

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/AdaptivePollingRecordPublisher.java (hunk 8 above)
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/AdaptivePollingRecordPublisher.java:36]: Thanks! :D

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/model/StartingPosition.java (hunk 9 above)
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/model/StartingPosition.java:39]: Done

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordPublisher.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+	/**
+	 * Run the record publisher. Records will be consumed from the stream and published to the consumer.
+	 * The number of batches retrieved by a single invocation will vary based on implementation.
+	 *
+	 * @param startingPosition the position in the stream from which to consume
+	 * @param recordConsumer the record consumer in which to output records
+	 * @return a status enum to represent whether a shard has been fully consumed
+	 * @throws InterruptedException
+	 */
+	RecordPublisherRunResult run(StartingPosition startingPosition, Consumer<RecordBatch> recordConsumer) throws InterruptedException;
```
[the remark this answers is not in the document]
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordPublisher.java:40]: I need to double check this, I will follow up tomorrow. Essentially it was originally like this for EFO. As the EFO subscription passes records to the consumer in a callback it is still in the network thread. This means `ReadTimeout` can occur and the consumer could fail mid batch (If you apply very large backpressure). The `FanOutRecordPublisher` would then not know where to start consumption from, therefore the `ShardConsumer` was passing that in. However, I ended up splitting the network and shard consumer thread with a `BlockingQueue` in the end, so I can probably track the state in the `RecordPublisher` as you suggest.

I will investigate and get back to you.

**Comment by dannycranmer** [comment]
@tzulitai I have addressed a few of the comments, I will answer and address the rest tomorrow. Thanks!

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java (hunk 5 above)
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java:145]: See other comment regarding `StartingPosition`

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 14]
```diff
@@ -53,283 +49,111 @@
...
-			return getShardIteratorForSentinel(sequenceNumber);
+			startingPosition = StartingPosition.fromTimestamp(initTimestamp);
 		} else {
-			// we will be starting from an actual sequence number (due to restore from failure).
-			return getShardIteratorForRealSequenceNumber(sequenceNumber);
+			startingPosition = StartingPosition.restartFromSequenceNumber(checkNotNull(lastSequenceNum));
 		}
 	}
 
-	protected String getShardIteratorForSentinel(SequenceNumber sentinelSequenceNumber) throws InterruptedException {
```
[the remark this answers is not in the document]
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java:159]: 👍

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/PollingRecordPublisher.java [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+	 *
+	 * @return shard iterator
+	 */
+	@Nullable
+	private String getShardIterator(final StartingPosition startingPosition) throws InterruptedException {
+		if (startingPosition.getShardIteratorType() == LATEST && subscribedShard.isClosed()) {
+			return null;
+		}
+
+		return kinesisProxy.getShardIterator(
```
[the remark this answers is not in the document]
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/PollingRecordPublisher.java:161]: For timestamp the AWS client throws an exception if `time > now`. If `time <= now` it returns `null` if the shard is complete, otherwise the iterator.

If a sequence number is non-existent then the service throws a 400 client error and the application terminates. If the sequence number is invalid then the AWS client throws an exception.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordBatch.java (hunk 7 above)
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordBatch.java:48]: Done

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java (hunk 5 above)
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: It does not anymore.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordPublisher.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordPublisher.java]: @tzulitai I have refactored the code, turned out to be a substantial change to move 1 field!
- `RecordPublisher` now tracks position in stream, `StartingPosition` is only used at startup passed by the factory
- `ShardConsumer` now passes back the last consumed sequence number in the callback. This is required to allow the `PollingRecordPublisher` to recover in event of `ExpiredIterator`
- `ShardConsumer` still needs to track the last consumed sequence number for deaggregate filtering

Let me know what you think. 👍

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 16]
```diff
@@ -53,283 +49,111 @@
  */
 @Internal
 public class ShardConsumer<T> implements Runnable {
```
[the remark this answers is not in the document]
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java:53]: Done

**Comment by dannycranmer** [comment]
@tzulitai I have addressed all the comments and updated the EFO PR to the new `RecordPublisher` interface. Thanks.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/PollingRecordPublisher.java [hunk 17]
```diff
@@ -85,12 +88,8 @@
 
 		Preconditions.checkArgument(expiredIteratorBackoffMillis >= 0);
 		Preconditions.checkArgument(maxNumberOfRecordsPerFetch > 0);
-	}
 
-	@Override
-	public void initialize(StartingPosition startingPosition) throws InterruptedException {
-		nextStartingPosition = startingPosition;
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/PollingRecordPublisher.java]: With the `intialize` method now removed, could you also remove the preconditions check on line 101?
`nextStartingPosition` should now never be null.
Also, can `nextStartingPosition` now be a final field?

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordPublisherFactory.java [hunk 18]
```diff
@@ -37,6 +46,41 @@
...
+			StreamShardHandle streamShardHandle) throws InterruptedException;
+
+	/**
+	 * Determines the starting position in which the {@link RecordPublisher} should start consuming from.
+	 *
+	 * @param sequenceNumber the sequence number to start from
+	 * @param consumerConfig the consumer properties
+	 * @return the {@link StartingPosition} in which to start consuming from
+	 */
+	default StartingPosition getStartingPosition(
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordPublisherFactory.java:62]: Was there any particular reason to put this method in the `RecordPublisherFactory`, instead of letting the factory get a `StartingPosition` (and not a `SequenceNumber`) on `create`?
The conversion from sequence number to `StartingPosition` logic could probably be a separate static utility method somewhere that is unrelated to the actual implementation of the `RecordPublisherFactory`.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/PollingRecordPublisher.java (hunk 17 above)
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/PollingRecordPublisher.java]: Ah yes I missed that Precondition. I will remove that check. 
`nextStartingPosition` cannot be `final` as it is updated in `run` to recover in case of `ExpiredIteratorException`.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordPublisherFactory.java (hunk 18 above)
Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordPublisherFactory.java:62]: No reason in particular, I am happy to move it. I am thinking a good place could be in `StartingPosition` itself, as we already have `StartingPosition.restartFromSequenceNumber(SequenceNumber)`. I could add  `StartingPosition.restartFromSequenceNumber(SequenceNumber, Properties)`, although it does feel like that class should not know how to parse properties, it might be better than creating a new class for that single method. 

I will go with this and can change it if you disagree.

Comment (tzulitai) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordPublisherFactory.java:62]: That sounds good to me, thanks.
If we want to be extra explicit, perhaps only the required properties are parsed and then passed in to `StartingPosition.restartFromSequenceNumber`

Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordPublisherFactory.java:62]: 👍

Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/RecordPublisherFactory.java]: It didn't feel right to pass the `@Nullable` fields to `StartingPosition` so in the end:
- `Date` is parsed from properties in `KinesisConfigUtil`
- Logic to determine to restart from timestamp or sequence is in `AWSUtil`

**Comment by tzulitai** [comment]
LGTM, +1.
Thanks @dannycranmer, will merge this PR now after another CI run.
